src/System.Threading.Tasks.Channels/System/Runtime/CompilerServices/AutoResetAwaiter.cs

```diff
@@ -0,0 +1,155 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Runtime.CompilerServices
+{
+    /// <summary>Provides a reusable object that can be awaited by a consumer and manually completed by a producer.</summary>
+    /// <typeparam name="TResult">The type of data being passed from producer to consumer.</typeparam>
+    public sealed class AutoResetAwaiter<TResult> : IAwaiter<TResult>
```

Comment:
Public! I love it :). We've been looking for things like this. /cc @davidfowl 

Is there a ManualResetAwaiter? (He says, before finishing reviewing the code ;P)